File: app/src/main/java/io/sweers/catchup/model/medium/MediumPost.java
package io.sweers.catchup.model.medium;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

@AutoValue
public abstract class MediumPost {
  public static JsonAdapter<MediumPost> jsonAdapter(@NonNull Moshi moshi) {
    return new AutoValue_MediumPost.MoshiJsonAdapter(moshi);
  }

  public static Builder builder() {
    return new AutoValue_MediumPost.Builder();
  }

  @Nullable public abstract Collection collection();

  public abstract Post post();

  public abstract User user();

  @NonNull
  public String constructUrl() {
    return "https://medium.com/@" + user().username() + "/" + post().uniqueSlug();
  }

  @NonNull
  public String constructCommentsUrl() {
    return constructUrl() + "#--responses";
  }

  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder collection(@Nullable Collection value);

    public abstract Builder post(Post value);

    public abstract Builder user(User value);

    public abstract MediumPost build();
  }
}

File: services/medium/src/main/kotlin/io/sweers/catchup/service/medium/model/MediumPost.kt
/*
 * Copyright (c) 2018 Zac Sweers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.sweers.catchup.service.medium.model

import com.squareup.moshi.JsonClass
import io.sweers.catchup.service.api.HasStableId

@JsonClass(generateAdapter = true)
internal data class MediumPost(val collection: Collection?,
    val post: Post,
    val user: User) : HasStableId {

  fun constructUrl() = "https://medium.com/@${user.username}/${post.uniqueSlug}"

  fun constructCommentsUrl() = "${constructUrl()}#--responses"

  override fun stableId(): Long = post.id.hashCode().toLong()

}
